org.eclipse.xtext.builder/src/org/eclipse/xtext/builder/impl/XtextBuilder.java

```diff
@@ -495,6 +516,8 @@ protected boolean isOpened(IResourceDelta delta) {
 	protected void clean(IProgressMonitor monitor) throws CoreException {
 		SubMonitor progress = SubMonitor.convert(monitor, 10);
 		try {
+			pollQueuedBuildData();
```

Comment:
So this is the problematic call?

Reply:
yes. there is an unrelated project with deltas (on generated code)
this call leads to a needsRebuild.
the rebuild detects that the project that was cleaned has no lastcurrentbuildstate
and thus does an immediate fullbuild on the cleaned build

Reply:
If I understand this correctly, some code / callgraph lead to unrelated projects being in the SimpleProjectDependencyGraph?

Reply:
i dont see how simpleProjectdependencygraph is involved.
needsrebuild is globally.

Reply:
and org.eclipse.xtext.builder.impl.javasupport.JdtQueuedBuildData.doNeedRebuild(JavaBuilderState, Procedure1<? super UnconfirmedStructuralChangesDelta>)
does not care about any graphs

Reply:
Understood. The offending full-build trigger is not visible in this change set here but is 'legacy' from `JdtQueuedBuildData.doNeedRebuild` - we may need the simpleProjectdependencygraph and check, if the project without delta is even related to the "asking" project. Would that make sense?

Reply:
cant we simply skip the clean part?

Reply:
True